src/lib.rs

```diff
@@ -0,0 +1,534 @@
+                Ok(outcome) => format!("{}", outcome),
+                Err(e) => format!("{}", e),
+            };
+            writeln!(
+                f,
+                "{:<30} {}",
+                format!("Ran {:?}", report.test_type),
+                outcome
+            )?;
+        }
+        Ok(())
+    }
+}
+
+impl MemtestReportList {
+    pub fn iter(&self) -> std::slice::Iter<'_, MemtestReport> {
+        self.reports.iter()
+    }
+
+    /// Returns true if all tests were run successfully and all tests passed
+    pub fn all_pass(&self) -> bool {
+        self.iter()
+            .all(|report| matches!(report.outcome, Ok(MemtestOutcome::Pass)))
+    }
+}
+
+impl MemtestReport {
+    fn new(test_type: MemtestType, outcome: Result<MemtestOutcome, MemtestError>) -> MemtestReport {
+        MemtestReport { test_type, outcome }
+    }
+}
+
+impl TimeoutChecker {
+    fn new(start_time: Instant, timeout: Duration) -> TimeoutChecker {
+        TimeoutChecker {
+            deadline: start_time + timeout,
+            curr_test: None,
+        }
+    }
+
+    /// This function should be called in the beginning of a memtest.
+    /// It initializes struct members and set `expected_iter`.
+    fn init(&mut self, expected_iter: u64) {
+        self.curr_test = Some(TimeoutCheckerTestInstance {
+            test_start_time: Instant::now(),
+            expected_iter,
+            completed_iter: 0,
+            checkpoint: 1,
+            num_checks_completed: 0,
+            // TODO: Choice of starting interval is arbitrary for now.
+            checking_interval: Duration::from_nanos(1000),
+        });
+    }
+
+    // TODO: TimeoutChecker is quite intertwined with MemtestError, might need decoupling later
+    /// This function should be called in every iteration of a memtest
+    ///
+    /// To minimize overhead, the time is only checked at specific checkpoints.
+    /// The algorithm makes a prediction of how many iterations will be done during `checking_interval`
+    /// and sets the checkpoint based on the prediction.
+    ///
+    /// The algorithm also determines whether it is likely that the test will be completed
+    /// based on `work_progress` and `time_progress`.
+    /// If it is likely that the test will be completed, `checking_interval_ns` is scaled up to be more
+    /// lenient and reduce overhead.
+    fn check(&mut self) -> Result<(), MemtestError> {
+        let Some(test_instance) = &mut self.curr_test else {
+            panic!("Attempted to check timeout without initialization");
+        };
+        if test_instance.completed_iter < test_instance.checkpoint {
+            test_instance.completed_iter += 1;
+            return Ok(());
+        }
+
+        let curr_time = Instant::now();
+        if curr_time >= self.deadline {
+            return Err(MemtestError::Timeout);
+        }
+
+        let work_progress =
+            test_instance.completed_iter as f64 / test_instance.expected_iter as f64;
```

Comment:
I don't think there are ways to avoid the use of `as` for lossy conversion between int and float?

Reply:
You are correct :)